lib/cylc/xtrigger_mgr.py

```diff
@@ -0,0 +1,247 @@
+        self.clockx_map = {}
+        # When next to call a function, by signature.
+        self.t_next_call = {}
+        # Satisfied triggers and their function results, by signature.
+        self.sat_xtrig = {}
+        # Signatures of satisfied clock triggers.
+        self.sat_xclock = []
+        # Signatures of active functions (waiting on callback).
+        self.active = []
+        # All trigger and clock signatures in the current task pool.
+        self.all_xtrig = []
+        self.all_xclock = []
+
+        self.pflag = False
+
+        # For function arg templating.
+        self.farg_templ = {
+            TMPL_SUITE_NAME: suite,
+            TMPL_USER_NAME: user,
+            TMPL_SUITE_RUN_DIR: suite_run_dir,
+            TMPL_SUITE_SHARE_DIR: suite_share_dir,
+        }
+        self.broadcast_mgr = broadcast_mgr
+
+    def add_clock(self, label, fctx):
+        """Add a new clock xtrigger."""
+        self.clockx_map[label] = fctx
+
+    def add_trig(self, label, fctx):
+        """Add a new xtrigger."""
+        self.functx_map[label] = fctx
+        # Check any string templates in the function arg values (note this
+        # won't catch bad task-specific values - which are added dynamically).
+        for argv in fctx.kwargs.values():
+            for match in RE_STR_TMPL.findall(argv):
+                if match not in ARG_VAL_TEMPLATES:
+                    raise ValueError(
+                        "Illegal template in xtrigger %s: %s" % (label, match))
+
+    def load_xtrigger_for_restart(self, row_idx, row):
+        if row_idx == 0:
+            OUT.info("LOADING satisfied xtriggers")
+        sig, results_pickle = row
+        self.sat_xtrig[sig] = pickle.loads(results_pickle)
+
+    def housekeep(self):
+        """Delete satisfied xtriggers and xclocks no longer needed."""
+        rem = []
+        for sig in self.sat_xtrig:
+            if sig not in self.all_xtrig:
+                rem.append(sig)
+        for r in rem:
+            del self.sat_xtrig[r]
+        rem = []
+        for sig in self.sat_xclock:
+            if sig not in self.all_xclock:
+                rem.append(sig)
+        for r in rem:
+            self.sat_xclock.remove(r)
+
+    def satisfy_xclock(self, itask):
+        """Attempt to satisfy itask's clock trigger, if it has one."""
+        label, sig, ctx, satisfied = self._get_xclock(itask)
+        if satisfied:
+            return
+        if ctx.func(**ctx.kwargs):
+            satisfied = True
+            itask.state.xclock = (label, True)
+            self.sat_xclock.append(sig)
+            LOG.info('clock xtrigger satisfied: %s = %s' % (label, str(ctx)))
+
+    def _get_xclock(self, itask, sig_only=False):
+        label, satisfied = itask.state.xclock
+        ctx = deepcopy(self.clockx_map[label])
+        ctx.kwargs.update(
+            {
+                'point': str(itask.point),
+                'point_as_seconds': itask.get_point_as_seconds(),
```

Comment:
There might be a requirement, but point is provided (via string templating) and functions can therefore compute point_as_seconds if needed ... however, I suppose it might be better to provide it routinely.